OpenRA.Game/TraitDictionary.cs

```diff
@@ -43,23 +43,72 @@ class TraitDictionary
 		static readonly Func<Type, ITraitContainer> CreateTraitContainer = t =>
 			(ITraitContainer)typeof(TraitContainer<>).MakeGenericType(t).GetConstructor(Type.EmptyTypes).Invoke(null);
 
-		readonly Dictionary<Type, ITraitContainer> traits = new Dictionary<Type, ITraitContainer>();
+		private ITraitContainer[] traits = null;
+
+		public TraitDictionary() {
+			Resize();
+		}
+
+		private void Resize() {
+			// New trait types have been registered. Ensure our array is same size as trait type index map.
+			int newSize = TraitTypeIndexMap.GetCount() + 64;
+			if (traits == null) {
```

Comment:
However, I'm not sure if we want to enforce those rules before a proper review happened, as we could end up touching those changes anyway.